File: reader.py
#!/usr/bin/env python

import os, sys, json, ast
import pymongo
from flask import *
from bson.objectid import ObjectId

_version = "1.0.0b"

class Remote:
    def __init__(self):
        try:
            uri = os.environ["REMOTE_URI"]
            self.client = pymongo.MongoClient(uri)
            self.db = self.client[uri[uri.rfind('/') + 1:]]
            self.Posts = self.db.Posts
        except Exception as e:
            print "Unable to establish connection to remote. System returned: "
            print type(e)
            print e
            sys.exit(-1)

app = Flask(__name__)
remote = Remote()

@app.route("/", methods=["GET"])
def index():
    try:
        posts = []
        for res in remote.Posts.find().sort("date.created", -1).limit(10):
            res["_id"] = str(res["_id"])
            posts.append(json.dumps(res))
        for i, j in enumerate(posts):
            posts[i] = ast.literal_eval(j)

        count = remote.Posts.count()
        return render_template('index.html', posts=posts, count=count)
    except Exception as e:
        print type(e)
        print e
        return render_template('index.html', error="There was an error loading posts.")

'''TODO
@app.route("/posts")
def posts():
    try:
        if remote.Posts.count() < 1:
            return "404: Not Found\nNo entries."
        else:
            if request.args.get('offset') == None:
                offset = 0
            else:
                offset = int(request.args.get('offset'))
            if request.args.get('limit') == None:
                limit = 100
            else:
                limit = int(request.args.get('limit'))
            ret = []
            for res in remote.Posts.find().sort("date.created", -1).limit(25):
                res["_id"] = str(res["_id"])
                ret.append(json.dumps(res))
            for i, j in enumerate(ret):
                ret[i] = ast.literal_eval(j)
            return render_template('cryptoc1/posts.html', posts=ret)
    except Exception as e:
        print e
        return redirect(url_for('error'))
'''

@app.route("/post/<post_id>", methods=["GET"])
def post(post_id):
    try:
        res = remote.Posts.find_one({"_id": ObjectId(post_id)})
        res["_id"] = str(res["_id"])
        res = ast.literal_eval(json.dumps(res))
        return render_template('post.html', post=res)
    except Exception as e:
        print type(e)
        print e
        return render_template('post.html', error="There was an error loading post.")

'''TODO
@app.route("/tag/<tag>", methods=["GET"])
def tag():
    pass
'''

##
#
#   API REST Endpoints
#
##

def crossorigin(ret):
    resp = make_response(str(ret))
    resp.headers['Access-Control-Allow-Origin'] = "*"
    return resp

@app.route("/api/v1/posts", methods=["GET", "OPTIONS"])
def all_posts():
    '''
    Returns <limit> number of posts, after <offset>. Queries of such matter are always sorted by an object's date property.

    REST parameters:
        ?offset=<int>?limit=<int>
    '''
    if request.method == "GET":
        try:
            if remote.Posts.count() < 1:
                return "404: Not Found\nNo entries."
            else:
                if request.args.get('offset') == None:
                    offset = 0
                else:
                    offset = int(request.args.get('offset'))
                if request.args.get('limit') == None:
                    limit = 10
                else:
                    limit = int(request.args.get('limit'))
                ret = []
                for res in remote.Posts.find().sort("date.created", -1).skip(offset).limit(limit):
                    res["_id"] = str(res["_id"])
                    ret.append(json.dumps(res))
                for i, j in enumerate(ret):
                    ret[i] = ast.literal_eval(j)
                return crossorigin(ret)
        except Exception as e:
            print "Unable to establish connection to remote. System returned: "
            print type(e)
            print e
            return "500: Internal Server Error", 500
    else:
        return "405: Method Not Allowed", 405


@app.route("/api/v1/post/<post_id>", methods=["GET", "OPTIONS"])
def post_by_id(post_id):
    if request.method == "GET":
        if not remote.Posts.find_one({"_id": ObjectId(post_id)}):
            return "404: Not Found", 404
        else:
            if post_id == "" or post_id == None or post_id == "null" or len(post_id) < 24:
                return "400: Bad Request", 400
            else:
                res = remote.Posts.find_one({"_id": ObjectId(post_id)})
                res["_id"] = str(res["_id"])
                crossorigin(ret)
    else:
        return "405: Method Not Allowed", 405

if __name__ == "__main__":
    app.run(debug=True)

File: reader.js
var express = require('express'),
    handlebars = require('express-handlebars'),
    dotenv = require('dotenv'),
    app = express()

var mongo = require('mongodb'),
    MongoClient = mongo.MongoClient,
    ObjectID = mongo.ObjectID

var zipf = require('./lib/zipf.js')
zipf.init()

dotenv.config()

var url = process.env.REMOTE_URI || zipf.remote

app.use(express.static(zipf.staticPath))
app.use(express.static('public'))

app.set("views", zipf.themePath + "/views/")
app.engine('handlebars', handlebars({
    layoutsDir: zipf.themePath + "/views/layouts/",
    defaultLayout: 'default'
}))
app.set('view engine', 'handlebars')

app.get('/', function(req, res) {
    MongoClient.connect(url, function(err, db) {
        if (err) {
            res.status(522).render('index', Error(522, req.path))
            console.error(err)
        } else {
            var collection = db.collection('Posts')
            collection.find().sort({
                "date.created": -1
            }).limit(10).toArray(function(err, posts) {
                if (err) {
                    res.status(578).render('index', Error(578, req.path))
                    console.error(err)
                } else {
                    collection.count(function(err, count) {
                        if (err) {
                            console.error(err)
                        } else {
                            var showLoader = false
                            if (posts.length < count) {
                                showLoader = true
                            }
                            res.render('index', {
                                title: "Samuel Steele",
                                count: count,
                                posts: posts,
                                showLoader: showLoader
                            })
                        }
                        db.close()
                    })
                }
            })
        }
    })
})

app.get('/post/:id', function(req, res) {
    MongoClient.connect(url, function(err, db) {
        if (err) {
            res.status(522).render('post', Error(522, req.path))
            console.error(err)
        } else {
            if (!(/[a-f0-9]{24}/.test(req.params.id))) {
                res.status(433).render('post', Error(433, req.path))
            } else {
                db.collection('Posts').findOne({
                    _id: ObjectID(req.params.id)
                }, function(err, post) {
                    if (err) {
                        res.status(578).render('post', Error(578, req.path))
                        console.error(err)
                    } else {
                        // @TODO: Render tags
                        res.render('post', {
                            title: post.title,
                            post: post
                        })
                    }
                    db.close()
                })
            }
        }
    })
})

app.get('/resume', function(res, res) {
    res.render('resume')
})

app.get('/api/v1/posts', function(req, res) {
    MongoClient.connect(url, function(err, db) {
        if (err) {
            res.status(522).json(Error(522))
            console.error(err)
        } else {
            var offset = 0
            if (req.query.offset) {
                offset = parseInt(req.query.offset)
            }
            db.collection('Posts').find().sort({
                "date.created": -1
            }).skip(offset).limit(10).toArray(function(err, posts) {
                if (err) {
                    res.status(578).json(Error(578, req.path))
                    console.error(err)
                } else {
                    res.json(posts)
                }
                db.close()
            })
        }
    })
})

app.get('/api/v1/post/:id', function(req, res) {
    MongoClient.connect(url, function(err, db) {
        if (err) {
            res.status(522).json(Error(522, req.path))
            console.error(err)
        } else {
            db.collection('Posts').findOne({
                _id: new ObjectID(req.params.id)
            }, function(err, post) {
                if (err) {
                    res.status(578).json(Error(578, req.path))
                    console.error(err)
                } else {
                    res.json(post)
                }
                db.close()
            })
        }
    })
})

function Error(code, title) {
    switch (code) {
        case 433:
            return {
                title: title,
                success: false,
                error: {
                    code: 433,
                    label: "Invalid Post ID",
                    body: "Supplied Post ID is not an ObjectID"
                }
            }
        case 522:
            return {
                title: title,
                success: false,
                error: {
                    code: 522,
                    label: "No Connection",
                    body: "Unable to establish connection to database"
                }
            }
        case 578:
            return {
                title: title,
                success: false,
                error: {
                    code: 578,
                    label: "Database Error",
                    body: "Unable to query the database"
                }
            }
        default:
            return {
                title: title,
                success: false,
                error: {
                    code: 500,
                    label: "Internal Server Error",
                    body: "Internal Server Error"
                }
            }
    }
}

var server = app.listen(process.env.PORT || 5000, function() {
    console.log("Listening at http://0.0.0.0:%d", server.address().port)
})
